### PR TITLE
Warn when executing Gradle inside an undefined build

### PR DESCRIPTION
**Pull request by lacasseio**
Fixes https://github.com/gradle/gradle/issues/6918

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and [CI build status](https://builds.gradle.org/project.html?projectId=Gradle_Check&tab=projectOverview&branch_Gradle_Check=lacasseio%2Fdeprecate-undefined-build)
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/core/src/integTest/groovy/org/gradle/api/UndefinedBuildExecutionIntegrationTest.groovy [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+    def setup() {
+        executer.inDirectory(undefinedBuildDirectory)
+        executer.ignoreMissingSettingsFile()
+    }
+
+    @Unroll
+    def "shows deprecation warning when executing #task task in undefined build"() {
+        when:
+        executer.expectDeprecationWarning()
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/api/UndefinedBuildExecutionIntegrationTest.groovy]: Please use the new `expectDeprecationWarning(String)` API.

Comment (lacasseio) [reply subprojects/core/src/integTest/groovy/org/gradle/api/UndefinedBuildExecutionIntegrationTest.groovy]: Done.

Comment (lptr) [reply subprojects/core/src/integTest/groovy/org/gradle/api/UndefinedBuildExecutionIntegrationTest.groovy]: Thanks!

File: subprojects/core/src/integTest/groovy/org/gradle/api/UndefinedBuildExecutionIntegrationTest.groovy [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+    @Unroll
+    def "shows deprecation warning when executing #task task in undefined build"() {
+        expect:
+        executer.expectDeprecationWarning("Executing Gradle tasks as part of an undefined build has been deprecated. This will fail with an error in Gradle 7.0.")
+        succeeds(task) // should fail for all tasks in Gradle 7.0
+        // In Gradle 7.0, we should test the .gradle folder is not created
+
+        where:
+        task << [ProjectInternal.HELP_TASK, ProjectInternal.PROJECTS_TASK, ProjectInternal.TASKS_TASK, HelpTasksPlugin.COMPONENTS_TASK, HelpTasksPlugin.DEPENDENCIES_TASK, HelpTasksPlugin.DEPENDENT_COMPONENTS_TASK, HelpTasksPlugin.MODEL_TASK, HelpTasksPlugin.OUTGOING_VARIANTS_TASK, HelpTasksPlugin.PROPERTIES_TASK]
```
Comment (big-guy) [inline subprojects/core/src/integTest/groovy/org/gradle/api/UndefinedBuildExecutionIntegrationTest.groovy]: I think just running a single task here is OK.  There's nothing special about any of these tasks.

File: subprojects/core/src/integTest/groovy/org/gradle/api/UndefinedBuildExecutionIntegrationTest.groovy [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    def "does not treat buildSrc as undefined build"() {
+        given:
+        settingsFile.touch()
+        file("buildSrc").mkdir()
+
+        expect:
+        succeeds("help") // without deprecation warning
```
Comment (big-guy) [inline subprojects/core/src/integTest/groovy/org/gradle/api/UndefinedBuildExecutionIntegrationTest.groovy]: Should this check for buildSrc tasks being executed?

File: subprojects/core/src/integTest/groovy/org/gradle/api/UndefinedBuildExecutionIntegrationTest.groovy [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import org.gradle.test.fixtures.file.TestFile
+import spock.lang.Unroll
+
+import java.nio.file.Files
+
+class UndefinedBuildExecutionIntegrationTest extends AbstractIntegrationSpec {
+    def undefinedBuildDirectory = Files.createTempDirectory("gradle").toFile()
+
+    def setup() {
+        executer.inDirectory(undefinedBuildDirectory)
```
Comment (big-guy) [inline subprojects/core/src/integTest/groovy/org/gradle/api/UndefinedBuildExecutionIntegrationTest.groovy]: Should we check that there aren't any settings files above this one?

File: subprojects/core/src/main/java/org/gradle/internal/service/scopes/GradleScopeServices.java [hunk 5]
```diff
@@ -152,13 +153,14 @@ CommandLineTaskParser createCommandLineTaskParser(OptionReader optionReader, Tas
...
 
     BuildWorkExecutor createBuildExecuter(StyledTextOutputFactory textOutputFactory, IncludedBuildControllers includedBuildControllers, BuildOperationExecutor buildOperationExecutor) {
-        return new BuildOperationFiringBuildWorkerExecutor(
-            new IncludedBuildLifecycleBuildWorkExecutor(
-                new DefaultBuildWorkExecutor(
-                    asList(new DryRunBuildExecutionAction(textOutputFactory),
-                        new SelectedTaskExecutionAction())),
-                includedBuildControllers),
-            buildOperationExecutor);
+        return new DeprecateUndefinedBuildWorkExecutor(
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/internal/service/scopes/GradleScopeServices.java]: I would keep the build operation as the outer most executor unless this breaks something.  That way the deprecation will be associated with the outermost build operation

File: subprojects/docs/src/docs/userguide/upgrading_version_5.adoc [hunk 6]
```diff
@@ -271,6 +271,12 @@ However, selecting the `default` configuration yields the same result as before.
 Only the `runtime` configuration now contains more information which corresponds to the _runtimeElements_ variant of a Java library.
 In general, users are advised to migrate from `ivy.xml` to the new Gradle Module Metadata format.
 
+==== Executing Gradle inside an undefined build has been deprecated
+
+A Gradle build is defined by the `settings.gradle[.kts]` file in the current or any parent directories.
```
Comment (big-guy) [inline subprojects/docs/src/docs/userguide/upgrading_version_5.adoc]:
```suggestion
A Gradle build is defined by a `settings.gradle[.kts]` file in the current or a parent directory. Without a settings file, a Gradle build is undefined and will emit a deprecation warning. 

In Gradle 7.0, Gradle will only allow you to invoke the `init` task or diagnostic command line flags, such as `--version`, with undefined builds.
```

File: subprojects/docs/src/docs/userguide/upgrading_version_5.adoc [hunk 7]
```diff
@@ -271,6 +271,12 @@ However, selecting the `default` configuration yields the same result as before.
...
+Failing to meet this requirement, Gradle is facing an undefined build.
```
Comment (big-guy) [inline subprojects/docs/src/docs/userguide/upgrading_version_5.adoc]:
```suggestion
```

File: subprojects/docs/src/docs/userguide/upgrading_version_5.adoc [hunk 8]
```diff
@@ -271,6 +271,12 @@ However, selecting the `default` configuration yields the same result as before.
...
+Executing Gradle in this scenario is only valid when invoking the `init` task, requesting help (either via `--help` or `help` task only for `init` task), or any diagnostic command line flags such as `--version`.
```
Comment (big-guy) [inline subprojects/docs/src/docs/userguide/upgrading_version_5.adoc]:
```suggestion
```

File: subprojects/docs/src/docs/userguide/upgrading_version_5.adoc [hunk 9]
```diff
@@ -271,6 +271,12 @@ However, selecting the `default` configuration yields the same result as before.
 Only the `runtime` configuration now contains more information which corresponds to the _runtimeElements_ variant of a Java library.
 In general, users are advised to migrate from `ivy.xml` to the new Gradle Module Metadata format.
 
+==== Executing Gradle inside an undefined build has been deprecated
```
Comment (big-guy) [inline subprojects/docs/src/docs/userguide/upgrading_version_5.adoc]:
```suggestion
==== Executing Gradle without a settings file has been deprecated
```

**Pull request merged**